### ISSUE TITLE
The timeout of  exec probe did not take effect as expected.

### ISSUE THREAD

**Issue by NoicFank** [kubernetes/kubernetes#107306]
### What happened?

I want to use the `exec` probe, and I found the `timeoutSeconds` with exec probe did not take effect as expected.
here is the situation, let's take readinessProbe for example to show the question. 

I use `exec` probe in  `readinessProbe` and set the `timeoutSeconds` of  `readinessProbe` to 5, where I want the execution to return in 5 seconds.  Then, I set the `command` the sleep 10s, 100s, 150s, which corresponding to pod  `readiness-10s, readiness-100s, readiness-150s`.

Then, as expect, all the readinessProbe of three pods should failed, since the `timeoutSeconds` is 5. But, in fact, only `readiness-150s` was failed.

the yaml of pod is as below.
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: readiness-150s
spec:
  containers:
  - name: liveness
    image: alpine
    args:
    - /bin/sh
    - -c
    - touch /tmp/healthy; sleep 6000
    readinessProbe:
      exec:
        command:
        - /bin/sh
        - -c
        - sleep 150; cat /tmp/healthy
      initialDelaySeconds: 5
      periodSeconds: 5
      timeoutSeconds: 5
```

all the three pods should not ready as expect, but pods `readiness-10s` and `readiness-100s` were ready.
```
> $ kubectl get pods
NAME             READY   STATUS    RESTARTS   AGE
readiness-100s   1/1     Running   0          33m
readiness-10s    1/1     Running   0          34m
readiness-150s   0/1     Running   0          33m

```

```
> $ kubectl describe pods readiness-150s
...
Events:
  Type     Reason     Age                  From               Message
  ----     ------     ----                 ----               -------
  Normal   Scheduled  35m                  default-scheduler  Successfully assigned default/readiness-150s to 10.0.0.9
  Normal   Pulling    35m                  kubelet            Pulling image "alpine"
  Normal   Pulled     35m                  kubelet            Successfully pulled image "alpine"
  Normal   Created    35m                  kubelet            Created container liveness
  Normal   Started    35m                  kubelet            Started container liveness
  Warning  Unhealthy  4m20s (x5 over 29m)  kubelet            Readiness probe errored: rpc error: code = DeadlineExceeded desc = context deadline exceeded

```
the readinessProbe of pod `readiness-150s` was context timeout.

### What did you expect to happen?

As shown in the example above, 
all the readiness of three pods `readiness-10s, readiness-100s, readiness-150s` should failed.

### How can we reproduce it (as minimally and precisely as possible)?

apply the yaml above, ad edit the `exec.command` to sleep `10s, 100s, 150s`

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.3", GitCommit:"ca643a4d1f7bfe34773c74f79527be4afd95bf39", GitTreeState:"clean", BuildDate:"2021-07-15T21:04:39Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.4-tke.14", GitCommit:"c93cd0e4cfd27c8216a9ee2b6f1969a9d1294c5c", GitTreeState:"clean", BuildDate:"2021-08-18T11:09:02Z", GoVersion:"go1.15.10", Compiler:"gc", Platform:"linux/amd64"}
WARNING: version difference between client (1.21) and server (1.18) exceeds the supported minor version skew of +/-1

</details>

### Cloud provider

<details>
NONE
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```
Darwin ARCHELURONG-MB0 19.6.0 Darwin Kernel Version 19.6.0: Thu May  6 00:48:39 PDT 2021; root:xnu-6153.141.33~1/RELEASE_X86_64 x86_64

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by NoicFank** [comment]
@gjkim42  thanks for your reply. I found the reason, that there is a gate to control the probe timeout in docker. maybe the gate is close in my env.

kubernetes/pkg/kubelet/dockershim/exec.go
```
	if timeout > 0 && utilfeature.DefaultFeatureGate.Enabled(features.ExecProbeTimeout) {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, timeout)
		defer cancel()
	}
```

**Comment by gjkim42** [comment]
Or maybe this is the reason.

https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes

> Before Kubernetes 1.20, the field timeoutSeconds was not respected for exec probes: probes continued running indefinitely, even past their configured deadline, until a result was returned.

> This defect was corrected in Kubernetes v1.20. You may have been relying on the previous behavior, even without realizing it, as the default timeout is 1 second.

**Comment by NoicFank** [comment]
yeah, thanks，appreciate for your time.



发自我的iPhone


------------------ Original ------------------
From: Gunju Kim ***@***.***&gt;
Date: Tue,Jan 4,2022 10:12 PM
To: kubernetes/kubernetes ***@***.***&gt;
Cc: Dingzhu Lurong ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [kubernetes/kubernetes] The timeout of  exec probe did not take effect as expected. (Issue #107306)





 
Or maybe this is the reason.
 
https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes
  
Before Kubernetes 1.20, the field timeoutSeconds was not respected for exec probes: probes continued running indefinitely, even past their configured deadline, until a result was returned.
   
This defect was corrected in Kubernetes v1.20. You may have been relying on the previous behavior, even without realizing it, as the default timeout is 1 second.
  
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
